{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5c4fdb88-d7e7-4a23-af18-be6c3ca80c2a",
   "metadata": {},
   "source": [
    "# Concurrency\n",
    "\n",
    "### Loading Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "92ccdd8e-11c7-41e3-a262-c1e6822b5bc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Math\n",
    "import math\n",
    "from math import hypot, factorial, sqrt, ceil\n",
    "\n",
    "# Numerical Computing\n",
    "import numpy as np\n",
    "\n",
    "# Data Manipulation\n",
    "import pandas as pd\n",
    "\n",
    "# Data Visualization\n",
    "import seaborn\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Loggers\n",
    "import logging\n",
    "import logging.handlers\n",
    "\n",
    "# SQLite\n",
    "import sqlite3\n",
    "\n",
    "# Enum\n",
    "from enum import Enum, auto\n",
    "\n",
    "# Print\n",
    "from pprint import pprint\n",
    "\n",
    "# OS\n",
    "import io\n",
    "import re\n",
    "import sys\n",
    "import abc\n",
    "import csv\n",
    "import time\n",
    "import gzip\n",
    "import queue\n",
    "import heapq\n",
    "import socket\n",
    "import string\n",
    "import random\n",
    "import bisect\n",
    "import operator\n",
    "import datetime\n",
    "import contextlib\n",
    "import subprocess\n",
    "from decimal import Decimal\n",
    "from abc import ABC, abstractmethod\n",
    "\n",
    "# Types & Annotations\n",
    "import collections\n",
    "from __future__ import annotations\n",
    "from collections import defaultdict, Counter\n",
    "from collections.abc import Container, Mapping, Hashable\n",
    "from typing import TYPE_CHECKING\n",
    "from typing import Pattern, Match\n",
    "from typing import Hashable, Mapping, TypeVar, Any, overload, Union, Sequence, Dict, Deque, TextIO, Callable, ContextManager\n",
    "from typing import List, Protocol, NoReturn, Union, Set, Tuple, Optional, Iterable, Iterator, cast, NamedTuple, TYPE_CHECKING\n",
    "# from typing import \n",
    "\n",
    "# Functional Tools\n",
    "from functools import wraps, total_ordering, lru_cache\n",
    "\n",
    "# Files & Path\n",
    "import tarfile\n",
    "import logging\n",
    "import zipfile\n",
    "import fnmatch\n",
    "from pathlib import Path\n",
    "from urllib.request import urlopen\n",
    "from urllib.parse import urlparse\n",
    "\n",
    "# Dataclass\n",
    "from dataclasses import dataclass, field"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72711281-b203-4570-bb00-27b2f847a817",
   "metadata": {},
   "source": [
    "### Threads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e54044fa-9a27-4000-9c73-6d703f2c5066",
   "metadata": {},
   "outputs": [],
   "source": [
    "from threading import Thread, Lock"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "82c3598f-6caa-443a-8c9f-8144331b4e5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Chef(Thread):\n",
    "    def __init__(self, name: str) -> None:\n",
    "        super().__init__(name=name)\n",
    "        self.total = 0\n",
    "\n",
    "    def get_order(self) -> None:\n",
    "        self.order = THE_ORDERS.pop(0)\n",
    "\n",
    "    def prepare(self) -> None:\n",
    "        \"\"\"Simulate doing a lot of work with a BIG computation\"\"\"\n",
    "        start = time.monotonic()\n",
    "        target = start + 1 + random.random()\n",
    "        for i in range(1_000_000_000):\n",
    "            self.total += math.factorial(i)\n",
    "            if time.monotonic() >= target:\n",
    "                break\n",
    "        print(f\"{time.monotonic():.3f} {self.name} made {self.order}\")\n",
    "\n",
    "    def run(self) -> None:\n",
    "        while True:\n",
    "            try:\n",
    "                self.get_order()\n",
    "                self.prepare()\n",
    "            except IndexError:\n",
    "                break  # No more orders"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ceb054f3-7ece-4957-aa2b-9e9b2b308fc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "THE_ORDERS = [\n",
    "    \"Reuben\",\n",
    "    \"Ham and Cheese\",\n",
    "    \"Monte Cristo\",\n",
    "    \"Tuna Melt\",\n",
    "    \"Cuban\",\n",
    "    \"Grilled Cheese\",\n",
    "    \"French Dip\",\n",
    "    \"BLT\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8695176a-b144-4c32-9b7e-228e667b4e73",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mo = Chef(\"Michael\")\n",
    "Constantine = Chef(\"Constantine\")\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    random.seed(42)\n",
    "    Mo.start()\n",
    "    Constantine.start()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ffaf111-55c8-4a40-a30c-a88ff4c36b17",
   "metadata": {},
   "source": [
    "### Multiprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9b031689-69ea-4cd0-9751-d93209256ed8",
   "metadata": {},
   "outputs": [],
   "source": [
    "from multiprocessing import Process, cpu_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2cc838c1-4936-424e-8061-97eb4cb45aab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "920385.335 Constantine made Ham and Cheese\n"
     ]
    }
   ],
   "source": [
    "class MuchCPU(Process):\n",
    "    def run(self) -> None:\n",
    "        print(f\"OS PID {os.getpid()}\")\n",
    "\n",
    "        s = sum(2 * i + 1 for i in range(100_000_000))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "baa87170-6082-4b49-938c-8dcc75c9a549",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "920385.940 Michael made Reuben\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread Thread-5:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/threading.py\", line 1045, in _bootstrap_inner\n",
      "Exception in thread Thread-6:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/threading.py\", line 1045, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/var/folders/dk/f9pbhbp52qxc0613mpbtx8lm0000gn/T/ipykernel_40033/243109373.py\", line 3, in run\n",
      "Exception in thread Thread-7:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/threading.py\", line 1045, in _bootstrap_inner\n",
      "Exception in thread Thread-8:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/threading.py\", line 1045, in _bootstrap_inner\n",
      "Exception in thread Thread-9:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/threading.py\", line 1045, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/var/folders/dk/f9pbhbp52qxc0613mpbtx8lm0000gn/T/ipykernel_40033/243109373.py\", line 3, in run\n",
      "Exception in thread Thread-10:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/threading.py\", line 1045, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/var/folders/dk/f9pbhbp52qxc0613mpbtx8lm0000gn/T/ipykernel_40033/243109373.py\", line 3, in run\n",
      "Exception in thread Thread-11:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/threading.py\", line 1045, in _bootstrap_inner\n",
      "NameError: name 'os' is not defined. Did you mean: 's'?\n",
      "    self.run()\n",
      "  File \"/var/folders/dk/f9pbhbp52qxc0613mpbtx8lm0000gn/T/ipykernel_40033/243109373.py\", line 3, in run\n",
      "    self.run()\n",
      "  File \"/var/folders/dk/f9pbhbp52qxc0613mpbtx8lm0000gn/T/ipykernel_40033/243109373.py\", line 3, in run\n",
      "NameError: name 'os' is not defined. Did you mean: 's'?\n",
      "Exception in thread Thread-12:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/threading.py\", line 1045, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/var/folders/dk/f9pbhbp52qxc0613mpbtx8lm0000gn/T/ipykernel_40033/243109373.py\", line 3, in run\n",
      "Exception in thread Thread-13:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/threading.py\", line 1045, in _bootstrap_inner\n",
      "NameError: name 'os' is not defined. Did you mean: 's'?\n",
      "NameError: name 'os' is not defined. Did you mean: 's'?\n",
      "NameError: name 'os' is not defined. Did you mean: 's'?\n",
      "    self.run()\n",
      "  File \"/var/folders/dk/f9pbhbp52qxc0613mpbtx8lm0000gn/T/ipykernel_40033/243109373.py\", line 3, in run\n",
      "NameError: name 'os' is not defined. Did you mean: 's'?\n",
      "Exception in thread Thread-14:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/threading.py\", line 1045, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/var/folders/dk/f9pbhbp52qxc0613mpbtx8lm0000gn/T/ipykernel_40033/243109373.py\", line 3, in run\n",
      "NameError: name 'os' is not defined. Did you mean: 's'?\n",
      "    self.run()\n",
      "  File \"/var/folders/dk/f9pbhbp52qxc0613mpbtx8lm0000gn/T/ipykernel_40033/243109373.py\", line 3, in run\n",
      "Exception in thread Thread-15:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/threading.py\", line 1045, in _bootstrap_inner\n",
      "Exception in thread Thread-16:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/threading.py\", line 1045, in _bootstrap_inner\n",
      "NameError: name 'os' is not defined. Did you mean: 's'?\n",
      "    self.run()\n",
      "  File \"/var/folders/dk/f9pbhbp52qxc0613mpbtx8lm0000gn/T/ipykernel_40033/243109373.py\", line 3, in run\n",
      "    self.run()\n",
      "  File \"/var/folders/dk/f9pbhbp52qxc0613mpbtx8lm0000gn/T/ipykernel_40033/243109373.py\", line 3, in run\n",
      "NameError: name 'os' is not defined. Did you mean: 's'?\n",
      "NameError: name 'os' is not defined. Did you mean: 's'?\n",
      "    self.run()\n",
      "  File \"/var/folders/dk/f9pbhbp52qxc0613mpbtx8lm0000gn/T/ipykernel_40033/243109373.py\", line 3, in run\n",
      "NameError: name 'os' is not defined. Did you mean: 's'?\n",
      "NameError: name 'os' is not defined. Did you mean: 's'?\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "920386.635 Constantine made Monte Cristo\n",
      "920387.163 Michael made Tuna Melt\n",
      "work took 1.067 seconds\n"
     ]
    }
   ],
   "source": [
    "class MoreCPU(Thread):\n",
    "    def run(self) -> None:\n",
    "        print(f\"OS PID {os.getpid()}\")\n",
    "\n",
    "        s = sum(2 * i + 1 for i in range(100_000_000))\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    # workers = [MuchCPU() for f in range(cpu_count())]\n",
    "    workers = [MoreCPU() for f in range(cpu_count())]\n",
    "\n",
    "    t = time.perf_counter()\n",
    "    for p in workers:\n",
    "        p.start()\n",
    "    for p in workers:\n",
    "        p.join()\n",
    "    print(f\"work took {time.perf_counter() - t:.3f} seconds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c9950b3-676a-470c-8cf4-844c87e67e75",
   "metadata": {},
   "source": [
    "### Multiprocessing Pools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2f32f96e-c6d7-455b-858f-4c7e5ad5faf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from multiprocessing.pool import Pool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e59714d-90a3-4679-82f1-feed7727962c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "920388.375 Constantine made Cuban\n",
      "920388.847 Michael made Grilled Cheese\n",
      "920390.277 Constantine made French Dip\n",
      "920390.690 Michael made BLT"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Process SpawnPoolWorker-1:\n",
      "Process SpawnPoolWorker-2:\n",
      "Process SpawnPoolWorker-3:\n",
      "Process SpawnPoolWorker-4:\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-5:\n",
      "Process SpawnPoolWorker-6:\n",
      "Process SpawnPoolWorker-7:\n",
      "Process SpawnPoolWorker-8:\n",
      "Process SpawnPoolWorker-9:\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Process SpawnPoolWorker-10:\n",
      "Process SpawnPoolWorker-11:\n",
      "Process SpawnPoolWorker-12:\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Process SpawnPoolWorker-13:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-14:\n",
      "Traceback (most recent call last):\n",
      "Process SpawnPoolWorker-15:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-16:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-17:\n",
      "Traceback (most recent call last):\n",
      "Process SpawnPoolWorker-18:\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-19:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-20:\n",
      "Traceback (most recent call last):\n",
      "Process SpawnPoolWorker-21:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-22:\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-23:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-24:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-25:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-26:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-27:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-28:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-29:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-30:\n",
      "Traceback (most recent call last):\n",
      "Process SpawnPoolWorker-31:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-32:\n",
      "Traceback (most recent call last):\n",
      "Process SpawnPoolWorker-33:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-34:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-35:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-36:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-38:\n",
      "Process SpawnPoolWorker-37:\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-39:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-40:\n",
      "Traceback (most recent call last):\n",
      "Process SpawnPoolWorker-41:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-42:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-43:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-44:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-45:\n",
      "Traceback (most recent call last):\n",
      "Process SpawnPoolWorker-46:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-47:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n",
      "Process SpawnPoolWorker-48:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "           ^^^^^\n",
      "  File \"/Users/isisromero/anaconda3/envs/OOP/lib/python3.11/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: Can't get attribute 'prime_factors' on <module '__main__' (built-in)>\n"
     ]
    }
   ],
   "source": [
    "def prime_factors(value: int) -> list[int]:\n",
    "    \"\"\"\n",
    "    >>> set(prime_factors(42))\n",
    "    {2, 3, 7}\n",
    "    >>> set(prime_factors(97))\n",
    "    {97}\n",
    "    \"\"\"\n",
    "    if value in {2, 3}:\n",
    "        return [value]\n",
    "    factors: list[int] = []\n",
    "    for divisor in range(2, ceil(sqrt(value)) + 1):\n",
    "        quotient, remainder = divmod(value, divisor)\n",
    "        if not remainder:\n",
    "            factors.extend(prime_factors(divisor))\n",
    "            factors.extend(prime_factors(quotient))\n",
    "            break\n",
    "    else:\n",
    "        factors = [value]\n",
    "    return factors\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    to_factor = [random.randint(100_000_000, 1_000_000_000) for i in range(40_960)]\n",
    "    \n",
    "    with Pool() as pool:\n",
    "        results = pool.map(prime_factors, to_factor)\n",
    "    \n",
    "    primes = [\n",
    "        value for value, factor_list in zip(to_factor, results) if len(factor_list) == 1\n",
    "    ]\n",
    "    \n",
    "    print(f\"9-digit primes: {primes}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8578cbf-03d6-4747-a1ae-310564270018",
   "metadata": {},
   "source": [
    "### Queues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bc2dec8-7cf5-4009-bbdd-1d28cae37056",
   "metadata": {},
   "outputs": [],
   "source": [
    "if TYPE_CHECKING:\n",
    "    Query_Q = Queue[Union[str, None]]\n",
    "    Result_Q = Queue[List[str]]\n",
    "\n",
    "\n",
    "def search(paths: list[Path], query_q: Query_Q, results_q: Result_Q) -> None:\n",
    "    print(f\"PID: {os.getpid()}, paths {len(paths)}\")\n",
    "    lines: list[str] = []\n",
    "    for path in paths:\n",
    "        lines.extend(l.rstrip() for l in path.read_text().splitlines())\n",
    "\n",
    "    while True:\n",
    "        if (query_text := query_q.get()) is None:\n",
    "            break\n",
    "        results = [l for l in lines if query_text in l]\n",
    "        results_q.put(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ea3a7c4-4b0a-4b33-ad0a-045558768b1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "class DirectorySearch:\n",
    "    def __init__(self) -> None:\n",
    "        self.query_queues: list[Query_Q]\n",
    "        self.results_queue: Result_Q\n",
    "        self.search_workers: list[Process]\n",
    "\n",
    "    def setup_search(self, paths: list[Path], cpus: Optional[int] = None) -> None:\n",
    "        if cpus is None:\n",
    "            cpus = cpu_count()\n",
    "        worker_paths = [paths[i::cpus] for i in range(cpus)]\n",
    "        self.query_queues = [Queue() for p in range(cpus)]\n",
    "        self.results_queue = Queue()\n",
    "\n",
    "        self.search_workers = [\n",
    "            Process(target=search, args=(paths, q, self.results_queue))\n",
    "            for paths, q in zip(worker_paths, self.query_queues)\n",
    "        ]\n",
    "        for proc in self.search_workers:\n",
    "            proc.start()\n",
    "\n",
    "    def teardown_search(self) -> None:\n",
    "        # Signal process termination\n",
    "        for q in self.query_queues:\n",
    "            q.put(None)\n",
    "\n",
    "        for proc in self.search_workers:\n",
    "            proc.join()\n",
    "\n",
    "    def search(self, target: str) -> Iterator[str]:\n",
    "        print(f\"search queues={self.query_queues}\")\n",
    "        for q in self.query_queues:\n",
    "            q.put(target)\n",
    "\n",
    "        for i in range(len(self.query_queues)):\n",
    "            for match in self.results_queue.get():\n",
    "                yield match"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4193c178-53b4-41b9-b3d2-e145305c6980",
   "metadata": {},
   "outputs": [],
   "source": [
    "def all_source(path: Path, pattern: str) -> Iterator[Path]:\n",
    "    for root, dirs, files in os.walk(path):\n",
    "        for skip in {\".tox\", \".mypy_cache\", \"__pycache__\", \".idea\"}:\n",
    "            if skip in dirs:\n",
    "                dirs.remove(skip)\n",
    "        yield from (Path(root) / f for f in files if fnmatch(f, pattern))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f843561d-d1ff-49d9-9210-35241ca86d9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from multiprocessing import Process, Queue, cpu_count\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24c63d72-51c7-4d53-9a56-5f2b6c3a4bb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    ds = DirectorySearch()\n",
    "    base = Path.cwd().parent\n",
    "    all_paths = list(all_source(base, \"*.py\"))\n",
    "    ds.setup_search(all_paths)\n",
    "    for target in (\"import\", \"class\", \"def\"):\n",
    "        start = time.perf_counter()\n",
    "        count = 0\n",
    "        for line in ds.search(target):\n",
    "            # print(line)\n",
    "            count += 1\n",
    "        milliseconds = 1000 * (time.perf_counter() - start)\n",
    "        print(\n",
    "            f\"Found {count} {target!r} in {len(all_paths)} files \"\n",
    "            f\"in {milliseconds:.3f}ms\"\n",
    "        )\n",
    "    ds.teardown_search()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2a17670-abf1-4f83-b322-15a99d076dcc",
   "metadata": {},
   "source": [
    "## Futures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "754eda19-98cf-4a6b-897b-1336276ac654",
   "metadata": {},
   "outputs": [],
   "source": [
    "class ImportResult(NamedTuple):\n",
    "    path: Path\n",
    "    imports: set[str]\n",
    "\n",
    "    @property\n",
    "    def focus(self) -> bool:\n",
    "        return \"typing\" in self.imports\n",
    "\n",
    "\n",
    "class ImportVisitor(ast.NodeVisitor):\n",
    "    def __init__(self) -> None:\n",
    "        self.imports: set[str] = set()\n",
    "\n",
    "    def visit_Import(self, node: ast.Import) -> None:\n",
    "        # print(ast.dump(node))\n",
    "        for alias in node.names:\n",
    "            self.imports.add(alias.name)\n",
    "\n",
    "    def visit_ImportFrom(self, node: ast.ImportFrom) -> None:\n",
    "        # print(ast.dump(node))\n",
    "        if node.module:\n",
    "            self.imports.add(node.module)\n",
    "\n",
    "\n",
    "def find_imports(path: Path) -> ImportResult:\n",
    "    tree = ast.parse(path.read_text())\n",
    "    iv = ImportVisitor()\n",
    "    iv.visit(tree)\n",
    "    return ImportResult(path, iv.imports)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b159424c-2940-45ae-a815-1a94e6f39cab",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main(base: Path = Path.cwd()) -> None:\n",
    "    print(f\"\\n{base}\")\n",
    "    start = time.perf_counter()\n",
    "    with futures.ThreadPoolExecutor(24) as pool:\n",
    "        analyzers = [\n",
    "            pool.submit(find_imports, path) for path in all_source(base, \"*.py\")\n",
    "        ]\n",
    "        analyzed = (worker.result() for worker in futures.as_completed(analyzers))\n",
    "    for example in sorted(analyzed):\n",
    "        print(\n",
    "            f\"{'->' if example.focus else '':2s} \"\n",
    "            f\"{example.path.relative_to(base)} {example.imports}\"\n",
    "        )\n",
    "    end = time.perf_counter()\n",
    "    rate = 1000 * (end - start) / len(analyzers)\n",
    "    print(f\"Searched {len(analyzers)} files in {base} at {rate:.3f}ms/file\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1396d923-0dfc-47d5-807d-f0b2a06e1552",
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    options = get_options()\n",
    "    for path in options.path:\n",
    "        main(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db2f0462-f124-4676-9d1a-9a82a86b9b03",
   "metadata": {},
   "outputs": [],
   "source": [
    "## AsyncIO in Action"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
